### DISCUSSION TITLE
Getting 400 when RLS enabled, but SELECT is OK

### DISCUSSION THREAD

**Discussion by peachp** [supabase/supabase#7950] [Questions] [answered]
I have only one policy for a table called `event_line`. When I copy the USING expression from this policy and put it after `select * from event_line WHERE`, the query returns the expected data.

However, when I try to get the data via Supabase JS client, I get a 400 error in browser console. Disabling the RLS for this table gets rid of the error and I receive the expected data in my app, but this is not the desired solution of course.

To give you some background: we are a wedding agency and are building an app for our vendors and couples (clients). So a wedding = event, and an "event_line" is a service like cake or photography.

Here is the policy's USING expression, copied from Supabase Authentication > Policies:
```
(event IN ( SELECT event.id
   FROM event
  WHERE (EXISTS ( SELECT 1
           FROM event_line event_line_1
          WHERE ((event_line_1.event = event.id) AND ((event_line_1.vendor)::text IN ( SELECT profile.vendor
                   FROM profile
                  WHERE (profile.user_id = 'editedc3-f399-4d51-8abe-e21f0a0a0d9d'::uuid))))))))
```

And this is the SQL query I run via Supabase SQL Editor:
```
select * from event_line WHERE
(event IN ( SELECT event.id
   FROM event
  WHERE (EXISTS ( SELECT 1
           FROM event_line event_line_1
          WHERE ((event_line_1.event = event.id) AND ((event_line_1.vendor)::text IN ( SELECT profile.vendor
                   FROM profile
                  WHERE (profile.user_id = 'editedc3-f399-4d51-8abe-e21f0a0a0d9d'::uuid))))))))

```

As you can see the WHERE / USING conditions are identical.

This is the JS code to retrieve the data:
```
let { data, error, status } = await supabase
      .from('event')
      .select(`
        id,
        couple,
        destination,
        event_line(*)
      `)
      .limit(100)
```

Also there is a policy for the parent table `event`, as well as the `profile` table used in the policy, which seem to work: when I have RLS enabled on `event` but disabled on `event_line`, the above JS .select works fine.

Interestingly though, when:
`event`: RLS enabled
`event_line`: RLS enabled
...this doesn't work:
```
let { data, error, status } = await supabase
      .from('event')
      .select(`
        id,
        couple,
        destination
      `)
      .limit(100)
```
(note I excluded `event_line` from the select string)
That's probably because the policy for the `event` table also uses `event_line`...?

I'm using a hardcoded user id just to eliminate the possibility that I'm not using the auth.uid() correctly and of course because it won't work when running the SQL query :)
I'm of course logged in as the hardcoded user when running the JS client code.

Any help will be appreciated!

**Comment by GaryAustin1** [comment, marked as answer]
If I am reading your policy correctly, you are using event_line table in a query in the policy for event_line.  This won't work as an RLS policy as it causes recursion.  It works from SQL editor because it bypasses RLS.
You will need to put part of your policy in a security definer function (which bypasses RLS).
https://supabase.com/docs/guides/auth/row-level-security#policies-with-security-definer-functions

**Reply by peachp** [reply]
It worked, thanks so much!
